### PR TITLE
Added CLI command to export and import the configuration

### PR DESCRIPTION
**Pull request by ghostwheel42**
## What type of PR?

enhancement

## What does this PR do?

This PR adds a `config-dump` cli command and updates the `config-update` cli command to
handle all possible items and parameters.
This was done by adding generic **to_dict** and **from_dict** methods to the Base model, so it should be quite future-proof.
The changes to `config-update` are backwards-compatible to the old command.
I've only removed the undocumented yaml-section _managers_ - managers can now be defined in the _users_ section.

The YAML now looks like this:

```
  aliases:
    - email: email@example.com
      destination:
        - address@example.com
  
  domains:
    - name: example.com
      alternatives:
        - alternative.tld
  
  relays:
    - name: relay.example.com
      smtp: mx.example.com
  
  users:
    - email: postmaster@example.com
      displayed_name: 'Postmaster'
      enable_imap: true
      enable_pop: false
      enabled: true
      forward_destination:
        - address@remote.example.com
      forward_enabled: true
      forward_keep: true
      global_admin: true
      manager_of:
        - example.com
      password: '{BLF-CRYPT}$2b$12$...'
      spam_enabled: true
      spam_threshold: 80

```

### Related issue(s)
- Closes issue #1377

## Prerequistes
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [X] In case of feature or enhancement: documentation updated accordingly
- [X] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/guide.html#changelog) entry file.

### CODE REVIEW

File: docs/cli.rst [hunk 1]
```diff
@@ -62,60 +63,124 @@ primary difference with simple `user` command is that password is being imported
...
 
-without ``--delete-object`` option config-update will only add/update new values but will *not* remove any entries missing in provided YAML input.
+  relays:
+    - name: relay.example.com
+      comment: test
+      smtp: mx.example.com
 
-Users
------
+You can use ``--dry-run`` to test your YAML without omitting any changes to the database.
```
Comment (Diman0) [inline docs/cli.rst:117]: small typo. Change 'omitting' to 'commiting'

**Review by Diman0 — CHANGES_REQUESTED** [review]
This looks very interesting. Many users will be happy with the introduction of this functionality. Thank you for your effort!

I see you modified changelog.md. This is not how you report changelog entries. Please undo the changes to changelog.md.

See this link for how to create the changelog entries (https://mailu.io/1.7/contributors/guide.html#changelog). You create in the townscrier folder a file with the changelog line. When the software is released the build process adds these change log entries to the main change log file.

It might take a couple of weeks before we have fully reviewed this PR. We all do this in our spare time after all. Thank you again for this PR.

File: docs/cli.rst (hunk 1 above)
Comment (ghostwheel42) [reply docs/cli.rst]: thanks for spotting this

**Comment by ghostwheel42** [comment]
The tests fail because I've converted the --verbose option to a flag.
bors try

**Comment by Diman0** [comment]
bors try

**Comment by Nebukadneza** [comment]
bors try

File: docs/cli.rst [hunk 2]
```diff
@@ -62,60 +63,124 @@ primary difference with simple `user` command is that password is being imported
...
 config-update
 -------------
 
-The sole purpose of this command is for importing users/aliases in bulk and synchronizing DB entries with external YAML template:
+The purpose of this command is for importing domain-, relay-, alias- and user-configuration in bulk and synchronizing DB entries with an external YAML template.
 
 .. code-block:: bash
 
-  cat mail-config.yml | docker-compose exec -T admin flask mailu config-update --delete-objects
+  docker exec -i $(docker-compose ps -q admin) flask mailu config-update -nvd < mail-config.yml
```
Comment (Diman0) [inline docs/cli.rst:90]: Can you also document the short-hand parameters n, v , d?
You use -nvd in the example code without explaining what these parameters are.
n = --dry-run
v = --verbose
d = --delete-objects

**Review by Diman0 — CHANGES_REQUESTED** [review]
Please also document the parameters n v d for the config-dump in the documentation.

Although the syntax of the yml file created by config-dump is correct. It is not nicely formatted. The text is not indented of each section. This makes it difficulty to read.

how it is dumped now (the dots are spaces.):
aliases:
\- email: email@example.com
..comment: ''
..destination:
..\- address@example.com
..wildcard: false


How it looks more readable (and also used in the documentation)
aliases:
..- email: email@example.com
....comment: ''
....destination:
....- address@example.com
....wildcard: false

Is it possible to have more human readable indentation?

**Review by Diman0 — APPROVED** [review]
Looks good to me. Thank you for the additional changes. This is now not only great new functionality, but also functionality that is documented very well.  Thank you again for submitting his PR.

**Comment by ghostwheel42** [comment]
I've tested config import and export more thoroughly and found some areas for improvement:

1. **dkim key storage** - dkim keys are stored **on disk** and **on update** - this bypasses the "--dry-run" feature which uses SQLAlchemy's _commit_ and _rollback_ methods. This **bug** can be fixed by
  a) storing the keys in the database _and_ writing them to disk **on commit**
  b) storing the keys in redis _only_ and implement access from [rspamd directly via redis](https://rspamd.com/doc/modules/dkim_signing.html#dkim-keys-in-redis)
  c) storing the keys in the database _and_ copying them to redis **on commit**

2. **config-update vs. config-import** - when updating the configuration using "--delete-objects" the "config-update" is really a "config-import", because the option is "all or nothing". After importing a config.yml the configuration is _exactly_ what was defined in the yml (nothing more, nothing less).
  Currently the "config-update" without "--delete-objects" can _only_ create new objects and change attributes of existing objects, but it _can not_ delete objects or related objects. Deleting an attribute (ie resetting it to default) is also not possible.
 Related objects are today:
   - the destination of an alias
   - alternative names of a domain
   - forward destinations of an user
   - the domains an user is manager of
   - a user's authentication tokens
   - a user's fetchmail jobs

### My proposals:

**dkim key storage**: I think implementing a) would be the best first step. Moving to c) could be considered in the future.

**config-import**: Adding the command "config-import" would be easy. Removing "--delete-objects" could be done, but is not backwards compatible.

**config-update**: I'm aware that a sophisticated update mechanism should be implemented via **REST**, but I'd like to add the possibility to delete objects or reset attributes via yaml. This is my proposal - yaml is not really made for this, so it's not _pretty_, but _functional_:

  1. delete a specific object (`- -[PRIMARY_KEY]`):
```
users:
  - email: foo@example.com
    forward_destination:
      - -forward@example.com
```
  2. delete all (other) objects not in yaml (`- -*prune*-`):
```
users:
  - email: foo@example.com
    forward_destination:
      - added@example.com
      - -*prune*-
```
  3. reset an attribute to default (`value : -*reset*-`):
```
users:
  - email: foo@example.com
    enable_pop: -*reset*-
```

**REST API**: I would like to implement a REST API (#445) using Flask-RESTful after implementing the stuff above.

@Nebukadneza @Diman0 What's your opinion on this?

Alex

**Comment by Diman0** [comment]
Both ideas seem sensible to me. Im a bit tired now. Tomorrow when I have a clear head, I will have a more detailed look and will provide a proper response.

**Comment by Diman0** [comment]
For point 1 (dkim key storage - dkim keys are stored on disk and on update ) I also think that option A is best. This will have the least impact for other components. Introducing too many changes will make mailu too complex. 

For point 2 (config-update vs. config-import) your suggestions sound good. An config-import command better conveys what the command does (import and replace). With a command name like config-update, you do not expect to completely replace the config. 
This additional functionality must be documented. E.g. that prune only effects non specified objects on the same level (forward destinations only for user foo@example.com in your example). The documentation should also contain examples. 

If you want to discuss more or could use help, you can also reach out to us on our [matrix channel](https://matrix.to/#/#mailu:tedomum.net).

**Comment by Diman0** [comment]
@ghostwheel42 Once you are finished with the necessary changes for this PR, please add an explicit comment that the PR is ready for review. That way I know when I can review it.

**Comment by ghostwheel42** [comment]
@Diman0 I'm currently rewriting the import/export stuff in a separate branch using marshmallow for serialization. This can be re-used when implementing the REST API - my work-in-progress looks promising.
Once I'm done I'll update the import/export branch and this PR.

**Comment by ghostwheel42** [comment]
Hi @Diman0, I've finished the switch to marshmallow now - it's ready for review.
Sorry for the cosmetic changes (single quotes) I've introduced - I can undo this if you'd like.
The import is clean and easy, but the "update" mode was a pain to implement.

- I've not added pygments to the requirements, but when it is installed the output will be colorized.
- I've revived the old config-update action for backwards compatability. If you'd like I'll replace the old config-update by a call to "config-import -u" and also remove the "-u" from config-import.
- I've not touched the dkim-key storage yet. I think there should be a way to have multiple keys and select the active key (see  #1700). Multiple keys could also be stored _only_ on disk in a subfolder. the active key would be symlinked.

Alex

**Comment by Diman0** [comment]
Hello @ghostwheel42 . Thank you for pinging me. i hope to have time this weekend to review and test your PR.

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
Unfortunately the admin container fails to start with your changes. See the [failed travis build](https://travis-ci.org/github/Mailu/Mailu/builds/762471528?utm_source=github_status&utm_medium=notification).  I don't understand the error itself. It refers to YamlLexer defined in schemas.py. But I cannot locate it schemas.py anywhere. This might be one of the dependencies that is installed as part of building the image?
```
  File "/app/mailu/schemas.py", line 56, in <module>
    class MyYamlLexer(YamlLexer):
NameError: name 'YamlLexer' is not defined
```

Full error trace
```
Traceback (most recent call last):
  File "/usr/bin/flask", line 8, in <module>
    sys.exit(main())
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 894, in main
    cli.main(args=args, prog_name=name)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 557, in main
    return super(FlaskGroup, self).main(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 1132, in invoke
    cmd_name, cmd, args = self.resolve_command(ctx, args)
  File "/usr/lib/python3.8/site-packages/click/core.py", line 1171, in resolve_command
    cmd = self.get_command(ctx, cmd_name)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 515, in get_command
    rv = info.load_app().cli.get_command(ctx, name)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 372, in load_app
    app = locate_app(self, import_name, name)
  File "/usr/lib/python3.8/site-packages/flask/cli.py", line 235, in locate_app
    __import__(module_name)
  File "/app/mailu/__init__.py", line 7, in <module>
    from mailu import utils, debug, models, manage, configuration
  File "/app/mailu/manage.py", line 16, in <module>
    from mailu.schemas import MailuSchema, Logger, RenderJSON
  File "/app/mailu/schemas.py", line 56, in <module>
    class MyYamlLexer(YamlLexer):
NameError: name 'YamlLexer' is not defined
```

**Comment by ghostwheel42** [comment]
> It refers to YamlLexer defined in schemas.py. But I cannot locate it schemas.py anywhere.

It's imported from pygments - I'm using a local virtual env for (faster) testing and didn't notice.
I'd really like to colorize the output, but I don't know if you like the extra dependency just for color.
Alex

**Comment by Diman0** [comment]
I think it is fine to add this dependency. It makes the output so much better.  It is really an improvement for the readability.

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
bors try

**Comment by ghostwheel42** [comment]
bors try

**Comment by Diman0** [comment]
bors try

File: docs/cli.rst [hunk 3]
```diff
@@ -113,8 +115,197 @@ following are additional parameters that could be defined for users:
...
+      destination:
+        - user1@example.com
+        - user2@example.com
+
+  relay:
+    - name: relay.example.com
+      comment: test
+      smtp: mx.example.com
+
+config-update shows the number of created/modified/deleted objects after import.
```
Comment (Diman0) [inline docs/cli.rst:215]: I reckon config-update should be config-import

File: docs/cli.rst [hunk 4]
```diff
@@ -113,8 +115,197 @@ following are additional parameters that could be defined for users:
...
+To suppress all messages except error messages use ``--quiet``.
+By adding the ``--verbose`` switch the import gets more detailed and shows exactly what attributes changed.
+In all log messages plain-text secrets (dkim-keys, passwords) are hidden by default. Use ``--secrets`` to log secrets.
+If you want to test what would be done when importing without committing any changes, use ``--dry-run``.
+
+By default config-update replaces the whole configuration. ``--update`` allows to modify the existing configuration instead.
```
Comment (Diman0) [inline docs/cli.rst:221]: I think config-update should be config-import here as well.

**Review by Diman0 — CHANGES_REQUESTED** [review]
It looks good to me. It is difficult to review it for me since you use a lot advanced python constructs I'm not very familiar with. 

The smoke tests I performed all worked correctly. The colour output mode looks really nice.

**Review by Diman0 — APPROVED** [review]
Lgtm

**Comment by Diman0** [comment]
bors retry

**Review by Diman0 — APPROVED** [review]

**Comment by Diman0** [comment]
bors retry

**Review by Diman0 — APPROVED** [review]

**Pull request merged**